srml/support/src/storage/hashed/generator.rs

```diff
@@ -203,23 +208,64 @@ pub trait StorageValue<T: codec::Codec> {
 	///
 	/// `T` is required to implement `codec::EncodeAppend`.
 	fn append<S: HashedStorage<Twox128>, I: codec::Encode>(
-		items: &[I], storage: &mut S
-	) -> Result<(), &'static str> where T: codec::EncodeAppend<Item=I> {
+		items: &[I],
+		storage: &mut S,
+	) -> Result<(), &'static str>
+		where T: codec::EncodeAppend<Item=I>
+	{
 		let new_val = <T as codec::EncodeAppend>::append(
-			storage.get_raw(Self::key()).unwrap_or_default(),
+			// if the key exists, directly append to it.
+			storage.get_raw(Self::key()).unwrap_or_else(|| {
+				// otherwise, try and read a proper __provided__ default.
+				Self::Default::default().map(|v| v.encode())
+					// or just use the default value.
+					.unwrap_or_default()
+			}),
 			items,
 		).map_err(|_| "Could not append given item")?;
 		storage.put_raw(Self::key(), &new_val);
 		Ok(())
 	}
+
+	/// Safely append the given items to the value in the storage. If a codec error occurs, then the
+	/// old (presumably corrupt) value is replaced with the given `items`.
+	///
+	/// `T` is required to implement `codec::EncodeAppend`.
+	fn append_or_put<'a, S: HashedStorage<Twox128>, I>(
+		items: &'a[I],
+		storage: &mut S,
+	) where
+		T: codec::EncodeAppend<Item=I> + From<&'a[I]>,
+		I: 'a + codec::Encode + Clone,
+	{
+		Self::append(items, storage).unwrap_or_else(|_| Self::put(&items.clone().into(), storage));
+	}
+
+	/// Read the length of the value in a fast way, without decoding the entire value.
+	///
+	/// `T` is required to implement `Codec::DecodeLength`.
+	fn decode_len<S: HashedStorage<Twox128>>(storage: &mut S) -> Result<usize, &'static str>
+		where T: codec::DecodeLength, T: Len
+	{
+		// attempt to get the length directly.
+		if let Some(k) = storage.get_raw(Self::key()) {
+			<T as codec::DecodeLength>::len(&k).map_err(|e| e.what())
+		} else {
+			let default = Self::Default::default()
```

Comment:
The else branch should be
```
Ok(Self::Default::default().map(|v| v.len()).unwrap_or(0))
```

When we don't have default value, then the length is always `0`.

Reply:
I'm ok for that as well, my point is still that the perfect implementation would be that decode_len on an `Option<Vec>` returns an `Option<usize>` but I'm ok with that as well.

Reply:
I don't have a strong opinion on this. What would be the advantage to get a `None` as size?

Reply:
If both you don't have a strong opinion I would keep it as-is because of something like:

```rust
assert_eq(OptionVec::get(), None);
assert_eq(OptionVec::decode_len(), 0);
assert_eq(OptionVec::decode_len(), None);
```

Reply:
What? Your current solution does not reflect what you are saying in the last comment.

And `OptionVec::get() == None` implies `OptionVec::decode_len() ==0`.

I said that I don't have a strong opinion on if we want to follow @thiolliere approach or mine, but I don't want to follow the current implementation. Options having a default value does not makes any sense at all. I will come up with a PR that forbids that.

Reply:
> What? Your current solution does not reflect what you are saying in the last comment.

Does with the exception that it returns `Err(...)` instead of `None`. I see it aligned with _`perfect implementation would be that decode_len on an Option<Vec> returns an Option<usize>`_. 

But I see your bigger point here esp. with 

> options having a default value does not make any sense at all. I will come up with a PR that forbids that.

(fyi. I have some tests now that assign defaults to options. I assume they will be removed once they are banned. For now, they are good examples)

So will change as you recommended. thanks for the description 👍

Reply:
> I don't have a strong opinion on this. What would be the advantage to get a None as size?

It is just for logic for Some(vec) decode len returns Some(len) and for None decode len returns None.
Maybe sometimes someone want to know the len of 

> If both you don't have a strong opinion I would keep it as-is

it is more like there is three option:
```
assert_eq(OptionVec::get(), None);

// 1- Basti proposal:
assert_eq(OptionVec::decode_len(), Ok(0));
// 2- Current state:
assert_eq(OptionVec::decode_len(), Err(".."));
// 3- Thiolliere proposal but changing the signature might be cumbersome
assert_eq(OptionVec::decode_len(), Ok(None));
```

I'm ok with all solution actually, though yes 1- seems ambiguous to me and I prefer 2 or 3